### PR TITLE
feat: package all in one agent jar

### PR DESCRIPTION
**Pull request by YongwuHe**

### CODE REVIEW

File: arex-agent/src/main/java/io/arex/agent/JarUtil.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarInputStream;
+
+public class JarUtil {
+    private static final String JAR_SUFFIX = ".jar";
+    private static final String BOOTSTRAP_JAR = "bootstrap";
+    private static File bootstrapJarFile = null;
+    private static String runTimeJarFilePath = "";
+    public static void extractInnerJar(File file) {
```
Comment (mr3) [inline arex-agent/src/main/java/io/arex/agent/JarUtil.java:17]: Replace innerJar to nestedJar

File: arex-agent/src/main/java/io/arex/agent/JarUtil.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarInputStream;
+
+public class JarUtil {
```
Comment (mr3) [inline arex-agent/src/main/java/io/arex/agent/JarUtil.java:12]: JarUtils

File: arex-agent/src/main/java/io/arex/agent/JarUtil.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            throw new IOException("Entry is outside of the target dir: " + entry.getName());
+        }
+        try(InputStream inputStream = file.getInputStream(entry);
+            FileOutputStream outputStream = new FileOutputStream(outputFile)) {
+            byte[] buffer = new byte[1024];
+            int length;
+            while ((length = inputStream.read(buffer)) > 0) {
+                outputStream.write(buffer, 0, length);
+            }
+            if (entry.getName().contains(BOOTSTRAP_JAR)) {
```
Comment (mr3) [inline arex-agent/src/main/java/io/arex/agent/JarUtil.java:55]: `BOOTSTRAP_JAR` This should be used as a method parameter

File: arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java [hunk 4]
```diff
@@ -62,7 +69,7 @@ private void addJarToLoaderSearch(File file, boolean isExtensionJar) {
 
             } while (jarEntry != null);
         } catch (Throwable ex) {
-            System.err.printf("add jar classes to advice failed, file: %s%n", file.getAbsolutePath());
+            System.err.printf("add jar classes to advice failed, file: %s%n ex: %s", file.getAbsolutePath(), ex);
```
Comment (mr3) [inline arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java:72]: file: %s, exception: %s%n

File: arex-instrumentation-api/src/main/java/io/arex/inst/runtime/listener/EventProcessor.java [hunk 5]
```diff
@@ -25,6 +27,14 @@ public class EventProcessor {
     private static final String CLOCK_METHOD = "currentTimeMillis";
     public static final String EXCLUDE_MOCK_TYPE = "java.util.HashMap-java.lang.String,java.util.HashSet";
     private static final AtomicReference<InitializeEnum> INIT_DEPENDENCY = new AtomicReference<>(InitializeEnum.START);
+    private static boolean noJacksonDependency = false;
```
Comment (mr3) [inline arex-instrumentation-api/src/main/java/io/arex/inst/runtime/listener/EventProcessor.java:30]: boolean jacksonSerializerFound

File: arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/AgentInitializer.java [hunk 6]
```diff
@@ -10,7 +10,10 @@ public class AgentInitializer {
 
     private static ClassLoader classLoader;
 
-    public static void initialize(Instrumentation inst, File agentFile, String agentArgs)
+    /**
+     * @param parentClassLoader APP classloader
```
Comment (mr3) [inline arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/AgentInitializer.java:14]: Normally, the parentClassLoader should be ClassLoaders.AppClassLoader

File: arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java [hunk 7]
```diff
@@ -95,6 +102,70 @@ private void addClassToInjectorCache(String adviceClassName) {
...
+    public void addSpecificJarToLoaderSearch(String jarName) {
+        String thirdPartyJarFilePaths = System.getProperty(ConfigConstants.THIRD_PARTY_JAR_FILE_PATH);
+        if (StringUtil.isEmpty(thirdPartyJarFilePaths)) {
+            return;
+        }
+        String[] split = StringUtil.split(thirdPartyJarFilePaths, ';');
+        for (String thirdPartyJarFilePath : split) {
+            if (StringUtil.containsIgnoreCase(thirdPartyJarFilePath, jarName)) {
+                appendToClassLoaderSearch(Thread.currentThread().getContextClassLoader(), new File(thirdPartyJarFilePath));
+            }
```
Comment (mr3) [inline arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java:163]: The assembly dependencySet output directory can be `/third-parth/jackson`.

We can use the folder name as the mapping key, put `arex.third.parth.xxx.path` into the system properties will get it directly without filtering it, such as `arex.third.parth.jackson.path`

File: arex-agent/src/main/java/io/arex/agent/JarUtils.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * ex: arex-agent-bootstrap.jar, jackson.jar
+     */
+    private static String extractNestedJar(JarFile file, JarEntry entry, String entryName) throws IOException {
+        String tempDir = System.getProperty("java.io.tmpdir");
+        File outputFile = new File(tempDir + File.separator + entryName);
+        String canonicalDestinationPath = outputFile.getCanonicalPath();
+        if (!canonicalDestinationPath.startsWith(tempDir)) {
+            throw new IOException("Entry is outside of the target dir: " + entryName);
+        }
```
Comment (mr3) [inline arex-agent/src/main/java/io/arex/agent/JarUtils.java:55]: There is no need to throw an IOException here

**Review by mr3 — DISMISSED** [review]

File: arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java [hunk 9]
```diff
@@ -95,6 +107,63 @@ private void addClassToInjectorCache(String adviceClassName) {
         }
     }
 
+    private static void appendToClassLoaderSearch(ClassLoader classLoader, File jarFile) {
+        try {
+            Method addURL = Class.forName("java.net.URLClassLoader").getDeclaredMethod("addURL", URL.class);
+            addURL.setAccessible(true);
```
Comment (lucas-myx) [inline arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java:113]: suggest set static variable, reduce reflect time

File: arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java [hunk 10]
```diff
@@ -95,6 +107,63 @@ private void addClassToInjectorCache(String adviceClassName) {
...
+             */
+            ClassLoader urlClassLoader = ClassLoader.getSystemClassLoader();
+            if (!(urlClassLoader instanceof URLClassLoader)) {
+                try (URLClassLoader tempClassLoader = new URLClassLoader(new URL[] {jarFile.toURI().toURL()}, urlClassLoader)) {
+                    addURL.invoke(tempClassLoader, jarFile.toURI().toURL());
+                    appendToAppClassLoaderSearch(classLoader, jarFile);
+                }
+            } else {
+                addURL.invoke(urlClassLoader, jarFile.toURI().toURL());
+                appendToAppClassLoaderSearch(classLoader, jarFile);
```
Comment (lucas-myx) [inline arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java:131]: take `appendToAppClassLoaderSearch` out and make it a common method

File: arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java [hunk 11]
```diff
@@ -95,6 +107,63 @@ private void addClassToInjectorCache(String adviceClassName) {
...
+
+    /**
+     * append jar jdk.internal.loader.ClassLoaders.AppClassLoader
+     * if java >= 11 need add jvm option:--add-opens=java.base/jdk.internal.loader=ALL-UNNAMED
+     * @param classLoader
+     * @param jarFile
+     */
+    private static void appendToAppClassLoaderSearch(ClassLoader classLoader, File jarFile) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        Class<? extends ClassLoader> loaderClass = classLoader.getClass();
+        if (JdkUtils.isJdk11() && ConfigConstants.APP_CLASSLOADER_NAME.equalsIgnoreCase(loaderClass.getName())) {
```
Comment (lucas-myx) [inline arex-agent-bootstrap/src/main/java/io/arex/agent/bootstrap/util/AdviceClassesCollector.java:147]: JdkUtils.isJdk11() -> jdk >= 11
method name easy to confuse

File: arex-agent/src/main/java/io/arex/agent/JarUtils.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+            } while (jarEntry != null);
+        } catch (Exception ex) {
+            System.err.printf("extract nested Jar failed, file: %s%n", file.getAbsolutePath());
+        }
+        return jarFileMap;
+    }
+
+    /**
+     * ex: arex-agent-bootstrap.jar, jackson.jar
```
Comment (lucas-myx) [inline arex-agent/src/main/java/io/arex/agent/JarUtils.java:47]:
```
arex-agent.jar
| -- arex-agent-bootstrap.jar
| -- jackson.jar
      | -- jackson1.jar
      | -- jackson2.jar
```

**Review by mr3 — APPROVED** [review]

**Review by lucas-myx — APPROVED** [review]

**Pull request merged**